src/sort.rs

```diff
@@ -0,0 +1,264 @@
+    /// returning a key of type `K` that implements `Ord`.
+    ///
+    /// I don't know how to use this method.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ego_tree::tree;
+    /// let mut tree = tree!('a' => { 'd', 'b', 'c' });
+    /// tree.root_mut()
+    ///     .sort_by_id_key(|id, value| id + *value as usize); // {1+100, 2+98, 3+99}
+    /// assert_eq!(
+    ///     vec![&'b', &'d', &'c'],
+    ///     tree.root()
+    ///         .children()
+    ///         .map(|n| n.value())
+    ///         .collect::<Vec<_>>(),
+    /// );
+    /// ```
+    pub fn sort_by_id_key<K, F>(&mut self, mut f: F)
+    where
+        F: FnMut(usize, &T) -> K,
+        K: Ord,
+    {
+        if self.has_children() {
+            let (unsorted, sorted) = self.sort_handler(|nodes| {
+                nodes.sort_by_key(|node| f(node.0.to_index(), node.1));
+            });
+            self.swap(unsorted, sorted);
+        }
+    }
+
+    /// Applies a sorting function to the children of the current node and returns their IDs
+    /// before and after sorting.
+    ///
+    /// This function takes a mutable closure `f` that sorts a vector of tuples,
+    /// where each tuple consists of a `NodeId` and a reference to the node's value `&T`.
+    ///
+    /// # Returns
+    ///
+    /// A tuple containing:
+    /// - `Vec<NodeId>`: The original order of the children's `NodeId`s before sorting.
+    /// - `Vec<NodeId>`: The order of the children's `NodeId`s after applying the sorting function.
+    fn sort_handler<F>(&mut self, mut f: F) -> (Vec<NodeId>, Vec<NodeId>)
+    where
+        F: FnMut(&mut Vec<(NodeId, &T)>),
+    {
+        let children = unsafe { self.tree.get_unchecked(self.id()).children() };
+        let (unsorted, mut nodes): (Vec<_>, Vec<_>) =
+            children.map(|n| (n.id(), (n.id(), n.value()))).unzip();
+        f(&mut nodes);
+        let sorted = nodes.into_iter().map(|(id, _)| id).collect::<Vec<_>>();
+        (unsorted, sorted)
+    }
+
+    /// Reorders the children of the current node to match the specified sorted order.
+    ///
+    /// This method takes two vectors of `NodeId`s: `unsorted`, which represents the original
+    /// order of the node's children, and `sorted`, which represents the desired order after sorting.
+    /// It swaps nodes in the tree such that their order in the tree matches the `sorted` vector.
+    ///
+    /// # Parameters
+    ///
+    /// - `unsorted`: A vector of `NodeId`s representing the original order of the node's children.
+    /// - `sorted`: A vector of `NodeId`s representing the desired order of the node's children.
+    ///
+    /// # Safety
+    ///
+    /// This function uses unsafe code to access and modify the tree nodes. Ensure that the node
+    /// indices are valid and that the tree structure remains consistent after the operation.
+    fn swap(&mut self, unsorted: Vec<NodeId>, sorted: Vec<NodeId>) {

```

Comment:
As for the approach to sorting: I wonder if it would not be much simpler to just produce a sorted list of ID and then repeatedly call `append_id`, i.e. something like

```rust
let mut children = {
  let this = unsafe { self.tree.get_unchecked(self.id()) };
  this.children().collect::<Vec<_>>()
};
children.sort_by(|lhs, rhs| f(lhs, rhs));
for child in children {
  self.append_id(child.id());
}
```

Reply:
Of course, you need to rebase on #40 for this to work reliably.

Reply:
I have just merged #40 so you can rebase on master

Reply:
The initial idea was this, but it would result in all elements having to be moved once.
For example: `['b', 'd', 'c']`, if only the sorted order is recorded, then move `b` to the end, then move `c` to the end, and then move `d` to the end.
So I changed the code to compare the positions before and after sorting. In the above example, only the step before inserting `c` into `d` needs to be performed.

Reply:
I might be more efficient, but it also adds more and unsafe code. The list of children itself should be limited in size in typical applications. So personally, I would prefer to go simple first and then optimize if someone actually measures any issues.

Reply:
IMHO, paying a (hopefully small) performance penalty but having less (and simpler) unsafe code is worth it

Reply:
Okay, I'll make the changes tonight along with the one below.

This code is indeed too complex. I thought about it all night yesterday and only found the bug after writing unit tests today.